node_modules/oae-util/lib/cassandra.js

```diff
@@ -742,17 +697,58 @@ var constructUpsertCQL = module.exports.constructUpsertCQL = function(cf, rowKey
 };
 
 /**
+ * Prepare a CQL query
+ *
+ * @param  {String}     query                       The CQL query to prepare
+ * @param  {Object}     opts                        A set of additional options
+ * @param  {Object}     opts.cache                  Whether the prepared query can be retrieved from or stored in the cache
+ * @param  {Function}   callback                    Standard callback function
+ * @param  {Object}     callback.err                An error that occurred, if any
+ * @param  {Object}     callback.preparedQuery      The prepared query instance
+ */
+var prepare = function(query, opts, callback) {
+    opts = opts || {};
+    if (opts.cache && cachedPreparedQueries[query]) {
+        return callback(null, cachedPreparedQueries[query]);
+    } else {
+        client.prepare(query, function(err, preparedQuery) {
+            if (err) {
+                log().error({
+                    'err': err,
+                    'query': query
+                }, 'Could not prepare a query');
+                return callback({'code': 500, 'msg': 'Could not prepare a query'});
+            }
+
+            // Cache the prepared statement only if:
+            //  - we're told the query is cache-able
+            //  - there are less than 10 arguments being passed in. This is a simple heuristic to
+            //    avoid leaking memory when someone builds a dynamic query but forgets to set the
+            //    `opts.cache` value to false
+            if (opts.cache && query.split('?').length < 10) {
+                cachedPreparedQueries[query] = preparedQuery;
+            }
+
+            return callback(null, preparedQuery);
+        });
+    }
+};
```

Comment:
How much performance benefit is prepared statements giving us? Do you think it's really necessary, in favour of maybe a more explicit way of registering prepared statements w/ the driver?

E.g., sometime in the future, exposing a function `Cassandra.prepare` for a statement, and then high-frequency queries such as authz and activity writes can just prepare their statements on init, and then there wouldn't be any other code or heuristics associated to it.